library/mps/reader.c

```diff
@@ -0,0 +1,626 @@
+                new_frag, (unsigned) new_frag_len );
+
+    if( new_frag == NULL )
+        RETURN( MBEDTLS_ERR_READER_INVALID_ARG );
+
+    for( size_t idx=0; idx < new_frag_len; idx++ )
+        TRACE( trace_comment, "Data[%u] = %02x", (unsigned) idx, (unsigned) new_frag[idx] );
+
+    /* Feeding is only possible if no fragment is currently being processed. */
+    frag = rd->frag;
+    if( frag != NULL )
+        RETURN( MBEDTLS_ERR_READER_UNEXPECTED_OPERATION );
+
+    acc = rd->acc;
+    if( acc != NULL )
+    {
+        size_t aa, ar;
+
+        ar = rd->acc_share.acc_remaining;
+        aa = rd->acc_avail;
+
+        copy_to_acc = ar;
+        if( copy_to_acc > new_frag_len )
+            copy_to_acc = new_frag_len;
+
+        acc += aa;
+        memcpy( acc, new_frag, copy_to_acc );
+
+        TRACE( trace_comment, "Copy new data of size %u of %u into accumulator at offset %u",
+                (unsigned) copy_to_acc, (unsigned) new_frag_len, (unsigned) aa );
+        for( size_t idx=0; idx < copy_to_acc; idx++ )
+        {
+            TRACE( trace_comment, "copy[%u] = %u", (unsigned) idx,
+                   (unsigned) new_frag[idx] );
+        }
+
+        /* Check if, with the new fragment, we have enough data. */
+        ar -= copy_to_acc;
+        if( ar > 0 )
+        {
+            /* Need more data */
+            aa += copy_to_acc;
+            rd->acc_share.acc_remaining = ar;
+            rd->acc_avail = aa;
+            RETURN( MBEDTLS_ERR_READER_NEED_MORE );
+        }
+
+        TRACE( trace_comment, "Enough data available to serve user request" );
+
+        rd->acc_share.frag_offset = aa;
+        aa += copy_to_acc;
+        rd->acc_avail = aa;
+    }
+    else
+    {
+        rd->acc_share.frag_offset = 0;
+    }
+
+    rd->frag = new_frag;
+    rd->frag_len = new_frag_len;
+    rd->commit = 0;
+    rd->end = 0;
+    RETURN( 0 );
+}
+
+
+int mbedtls_reader_get( mbedtls_reader *rd, size_t desired,
+                        unsigned char **buffer, size_t *buflen )
+{
+    unsigned char *frag, *acc;
+    size_t end, fo, fl, frag_fetched, frag_remaining;
+    TRACE_INIT( "reader_get %p, desired %u", rd, (unsigned) desired );
+
+    /* Check that the reader is in consuming mode. */
+    frag = rd->frag;
+    if( frag == NULL )
+        RETURN( MBEDTLS_ERR_READER_UNEXPECTED_OPERATION );
+
+    /* Use synthetic fragment offset 0 if no accumulator is used. */
+    acc = rd->acc;
+    if( acc == NULL )
+        fo = 0;
+    else
+        fo = rd->acc_share.frag_offset;
+
+    TRACE( trace_comment, "frag_off %u, end %u, acc_avail %d",
+            (unsigned) fo, (unsigned) rd->end,
+            acc == NULL ? -1 : (int) rd->acc_avail );
+
+    end = rd->end;
+    if( end < fo )
+    {
+        TRACE( trace_comment, "Serve the request from the accumulator" );
+        /* Serve the request from the accumulator */
+        if( fo - end < desired )
+        {
+            /* Illustration of supported and unsupported cases:
```

Comment:
The accumulator behaviour described in reader.c, mbedtls_reader_get, is, in my opinion, misplaced. It should either be moved to the design document (just as the reader state graph), or the reader state graph should be moved into the source code.

Reply:
I placed it in `reader.c` because it's an implementation detail. I agree, though, that the design document is currently lacking a description of the general approach, indicating when and how the switch from the accumulator to the fragment happens.